### ISSUE TITLE
[Question] APIRequestContext python, should I use page context or playwright context directly?

### ISSUE THREAD

**Issue by ltsuda** [microsoft/playwright#11484]
Hi guys,

I have an E2E automation using Playwright Python (pytest) for study, I'm wondering what's the best approach for the following question:

- I have a class ApplicationAPI that treats all the requests/responses to the Backend directly, so I can use that to create/delete resources in test setups/teardown. Today, I'm using the `requests` library, but I'm converting to use Playwright.

My questions are:

1. Should I have a `context: APIRequestContext` property in this class so I use the same context as the browser context? 
2. Or should I use `playwright.request.new_context()`, in this case I don't need to refactor all classes/tests that uses the ApplicationAPI class to send the `context` into it.
2.1. This new context will be finished only after the test run finishes or after the response arrived? I'm afraid that each request will create a new context and stay on memory/using resource which could cause performances issue due to the number of open contexts until the test run finishes.

**Comment by ltsuda** [comment]
Reading more about the `APIRequestContext`, I think I got my answers :)

The class ApplicationAPI should open the `sync_playwright()` python context to perform all the requests, so it will close the context after the request/all my treatments ends.

```python
from playwright.sync_api import sync_playwright

#... class body

  def send_request(self):
      with sync_playwright() as p:
          api_request_context = playwright.request.new_context(base_url="myurl")
          api_request_context.fetch(...)
        # ...
       return something
```

Does this sound correctly?

**Comment by mxschmitt** [comment]
Yes this is correct.

Usually it's recommend to use the global request context over the page/context api request context instance. This is because the latter will write the received cookies over the browser context cookie storage and this is often not needed (and slower).

We'll soon follow-up with more documentation about api testing in Python.

**Comment by mxschmitt** [comment]
Could you create a small repro without pytest which we are able to run? Feel free to create a new issue at https://github.com/microsoft/playwright-python

**Comment by ltsuda** [comment]
I got it working but I'm not sure if it is 100% the best way to do it.

Basically, in the `conftest.py`, I added the Session Fixture `Playwright` into my Function fixture. Then, pass it down until the `ApplicationAPI`, so it didn't needed to call ` with playwright() as p:` in it. I'm thinking that in my last comment, the code was trying to open a new `Playwright` instance, which was inside of the Session Fixture `Playwright` from pytest-playwright.

If you think it is still valid to create a small repro, I will try to do that today and create the issue in the python repository. I just don't know if will be possible to reproduce it without using pytest-playwright as I'm using the Session/Function fixtures (https://playwright.dev/python/docs/test-runners#fixtures)

```python
# application_api.py
from playwright.sync_api import (APIRequestContext, APIResponse, Error,
                                 sync_playwright, Playwright)
                                 
class ApplicationAPI:
    def __init__(..., playwright: Playwright):
        self.playwright = playwright
        ... continue the class properties

    def _login(self):
            self.headers["Tenant"] = self.tenant
            self.headers["Content-Type"] = "application/json"
            headers = self.headers
            headers["Authorization"] = "Basic {}".format(
                get_encoded_credentials(self.user, self.password))
            response = None

            try:
                 # uses Session Playwright instance coming through conftest
                context: APIRequestContext = self.playwright.request.new_context(
                    base_url=URL.SERVER_PROXY)
                response: APIResponse = context.get(
                    URL.API_TOKEN, headers=headers, fail_on_status_code=True)

            except Error as error:
                if self.fail_ok:
                    # LOG.warning(exception.__str__())
                    print(error.message)
                else:
                    # raise exceptions.StudioConductorError(
                    #     "Login failed with message: {}".format(exception.__str__())
                    # )
                    raise error

            code: str = str(response.status)
            response = response.text()
            if code.startswith("20"):
                content = json.loads(response)
                return code, content["value"]
            else:
                return code, response

    def send_request(self):
        code, response = self._login()
        if not code.startswith("20"):
            return code, response
        ... continue the method implementation


# utils.py
def set_getting_started_status(playwright: Playwright, username: str = Credentials.USERNAME, enabled=True):
    api = ApplicationAPI(method=HTTPVerb.POST, endpoint="/users/{}".format(username),
                       data={"show_getting_started": enabled}, playwright=playwright)
    api.send_request()
 
def authenticate(playwright: Playwright, page: Page, remove_getting_started=True):
    response = stage_request(page.context, "./auth/login", "POST")
    ... continue implemenation

    if remove_getting_started:
        set_getting_started_status(playwright, enabled=False)
    else:
        set_getting_started_status(playwright, enabled=True)

    return page

    
 # conftest.py
@pytest.fixture(scope="function")
def authenticated_page_without_wizard(page: Page, playwright: Playwright): # request Session Playwright fixture
    page = authenticate(playwright, page, remove_getting_started=True)
    page.goto("")

    yield page
    page.close()

# test.py
from conftest import authenticated_page_without_wizard

class TestGettingStarted:

    def test_getting_started_is_present_if_query(self, authenticated_page_without_wizard):
        main_page = MainPage(authenticated_page_without_wizard)
        main_page.page_title().wait_for()

```

**Comment by mxschmitt** [comment]
> Basically, in the conftest.py, I added the Session Fixture Playwright into my Function fixture. Then, pass it down until the ApplicationAPI, so it didn't needed to call with playwright() as p: in it.

this is amazing, thats the same way we are doing it in our official pytest plugin. ref: https://github.com/microsoft/playwright-pytest

No need for a repro then if it all works then as expected. 👍

**Comment by ltsuda** [comment]
Thanks for the support :))))

I kind realized looking here https://github.com/microsoft/playwright-python/blob/main/tests/sync/conftest.py
